### PR TITLE
Support flexible shapes on Oracle OCI

### PR DESCRIPTION
**Pull request by nvinuesa**
In the context of adding ARM support for Oracle OCI (https://github.com/juju/juju/pull/16277), we must fist support oracle flexible shapes because the only ARM shape on OCI is a flexible one (https://docs.oracle.com/en-us/iaas/Content/Compute/References/computeshapes.htm#flexible).

In order to support flexible (https://docs.oracle.com/en-us/iaas/Content/Compute/References/computeshapes.htm#flexible) shapes we must pass some configuration `ShapeConfig` (at least the number of cpu-cores) at the moment of launching the instance. This configuration value will be either the user-passed constraint(s) cpu-cores and memory, else a default minimum cpu-cores. We also define a new constant for minimum cpu-cores that is matched on `instances.MatchingInstanceTypes`. 

Also, we cannot pass this `ShapeConfig` value if the selected shape is not flexible so we have to be able to distinct between flexible and non-flexible shapes. This was done by adding `MaxCpuCores` and `MaxMem` on `InstanceType`

_Bonus: some cleanup on tests._

## Checklist

*If an item is not applicable, use `~strikethrough~`.*

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

```
juju bootstrap --debug --config compartment-id=ocid1.compartment.oc1..aaaaaaaanvu2racnlczevenu73dlcf3nokgsjpdkbdgp4xbrz3lb2y4giyjq oci-canonical c
```
check OCIs console and make sure the `VM.Standard1.1` shape was selected as the controller running instance.
 
## Links

**Jira card:** JUJU-[4694]

### CODE REVIEW

**Review by hmlanigan — COMMENTED** [review]
Mostly small things, I tested with google, aws and oracle. An oracle Flex shape was never chosen, but I did request one and it was used via constraints.

**Review by SimonRichardson — COMMENTED** [review]
We need a test somewhere that states that you can't set max cores without setting min cores. It should be an error or a warning at the very least.

File: provider/oci/provider.go [hunk 1]
```diff
@@ -41,27 +41,6 @@ type environConfig struct {
...
-// 				Type: []jsonschema.Type{jsonschema.ObjectType},
-// 			},
-// 		},
-// 		cloud.AuthTypesKey: {
-// 			// don't need a prompt, since there's only one choice.
-// 			Type: []jsonschema.Type{jsonschema.ArrayType},
-// 			Enum: []interface{}{[]string{string(cloud.HTTPSigAuthType)}},
-// 		},
-// 	},
-// }
```
Comment (SimonRichardson) [inline provider/oci/provider.go:63]: Delete me.

**Comment by nvinuesa** [comment]
@SimonRichardson 
> We need a test somewhere that states that you can't set max cores without setting min cores. It should be an error or a warning at the very least.

What do you mean? we are never setting min cores anywhere, we only have a new attribute `MaxCpuCores` and a constant `MinCpuCores`. 
You meant the `ShapeOcpuOptions.Min` value that is part of the OCI API's response?

File: provider/oci/provider.go (hunk 1 above)
Comment (nvinuesa) [reply provider/oci/provider.go:63]: these lines are deleted on this PR actually

Comment (SimonRichardson) [reply provider/oci/provider.go:63]: Ah yeah, whoops.

File: environs/instances/instancetype.go [hunk 2]
```diff
@@ -22,9 +22,11 @@ type InstanceType struct {
...
 	RootDisk   uint64
 	// These attributes are not supported by all clouds.
-	VirtType *string // The type of virtualisation used by the hypervisor, must match the image.
-	CpuPower *uint64
-	Tags     []string
+	VirtType    *string // The type of virtualisation used by the hypervisor, must match the image.
+	CpuPower    *uint64
+	Tags        []string
+	MaxCpuCores *uint64
+	MaxMem      *uint64
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply environs/instances/instancetype.go:32]: done

File: provider/oci/images.go [hunk 3]
```diff
@@ -325,6 +330,18 @@ func instanceTypes(cli ComputeClient, compartmentID, imageID *string) ([]instanc
 			CpuCores: uint64(cpus),
 			VirtType: &instanceType,
 		}
+		// If the shape is a flexible shape then the MemoryOptions and
+		// OcpuOptions will not be nil and they  indicate the maximum
+		// and minimum values. We assign the max memory and cpu cores
+		// values to the instance type in that case.
+		if val.MemoryOptions != nil {
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply provider/oci/images.go:337]: nice catch, done

File: provider/oci/images.go [hunk 4]
```diff
@@ -325,6 +330,18 @@ func instanceTypes(cli ComputeClient, compartmentID, imageID *string) ([]instanc
...
+			maxMem := uint64(*val.MemoryOptions.MaxInGBs) * 1024
+			newType.MaxMem = &maxMem
+		}
+		if val.OcpuOptions != nil {
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply provider/oci/images.go:343]: nice catch, done

File: provider/oci/images.go [hunk 5]
```diff
@@ -428,6 +445,7 @@ func refreshImageCache(cli ComputeClient, compartmentID *string) (*ImageCache, e
 	images := map[corebase.Base][]InstanceImage{}
 
 	for _, val := range items {
+		logger.Warningf("*** LISTING SHAPES FOR %s", val.String())
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply provider/oci/images.go]: I had forgotten to clean that up, removed

File: provider/oci/environ.go [hunk 6]
```diff
@@ -617,6 +617,25 @@ func (e *Environ) startInstance(
 		FreeformTags: tags,
 	}
 
+	// If the selected spec is a flexible shape, we must provide the number
+	// of OCPUs at least, so if the user hasn't provided cpu constraints we
+	// must pass the default value.
+	if (spec.InstanceType.MaxCpuCores != nil && spec.InstanceType.MaxCpuCores != &spec.InstanceType.CpuCores) ||
+		(spec.InstanceType.MaxMem != nil && spec.InstanceType.MaxMem != &spec.InstanceType.Mem) {
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply provider/oci/environ.go]: No, this checks each one of them independently with OR. But indeed from what I've seen empirically both values are always there.

File: provider/oci/environ.go [hunk 7]
```diff
@@ -617,6 +617,25 @@ func (e *Environ) startInstance(
...
+		instanceDetails.ShapeConfig = &ociCore.LaunchInstanceShapeConfigDetails{}
+		if args.Constraints.HasCpuCores() {
+			cpuCores := float32(*args.Constraints.CpuCores)
+			instanceDetails.ShapeConfig.Ocpus = &cpuCores
+		} else {
+			cpuCores := float32(instances.MinCpuCores)
+			instanceDetails.ShapeConfig.Ocpus = &cpuCores
+		}
+		if args.Constraints.HasMem() {
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply provider/oci/environ.go]: sure, done

File: provider/oci/images.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (hmlanigan) [reply provider/oci/images.go:337]: These two checks can be done in one line, if the first fails, the 2nd isn't evaluated. Same with the next.

**Comment by hmlanigan** [comment]
@nvinuesa do you plan to squash 270f559 before merging? The commit message is inaccurate.

**Review by hmlanigan — APPROVED** [review]
I'm happy if @SimonRichardson is

**Pull request merged**